cocotb/share/lib/embed/gpi_embed.c

```diff
@@ -243,121 +268,133 @@ int embed_sim_init(gpi_sim_info_t *info)
         goto cleanup;
 
     // Obtain the loggpi logger object
-    simlog_obj = PyObject_GetAttrString(cocotb_module, "loggpi");
+    simlog_obj = PyObject_GetAttrString(cocotb_module, "loggpi");       // New reference
 
     if (simlog_obj == NULL) {
         PyErr_Print();
         LOG_ERROR("Failed to get simlog object from loggpi\n");
+        goto cleanup;
     }
 
-    simlog_func = PyObject_GetAttrString(simlog_obj, "_logFromC");
-    if (simlog_func == NULL) {
+    // Obtain the function to use when logging from C code
+    log_func = PyObject_GetAttrString(simlog_obj, "_logFromC");         // New reference
+    if (log_func == NULL) {
         PyErr_Print();
         LOG_ERROR("Failed to get the _logFromC method");
         goto cleanup;
     }
 
-    if (!PyCallable_Check(simlog_func)) {
+    if (!PyCallable_Check(log_func)) {
         PyErr_Print();
         LOG_ERROR("_logFromC is not callable");
+        Py_DecRef(log_func);
         goto cleanup;
     }
 
-    set_log_handler(simlog_func);
-
-    Py_DECREF(simlog_func);
+    set_log_handler(log_func);      // Note: This function steals a reference to log_func.
 
-    simlog_func = PyObject_GetAttrString(simlog_obj, "isEnabledFor");
-    if (simlog_func == NULL) {
+    // Obtain the function to check whether to call log function
+    filter_func = PyObject_GetAttrString(simlog_obj, "isEnabledFor");   // New reference
+    if (filter_func == NULL) {
         PyErr_Print();
         LOG_ERROR("Failed to get the isEnabledFor method");
         goto cleanup;
     }
 
-    if (!PyCallable_Check(simlog_func)) {
+    if (!PyCallable_Check(filter_func)) {
         PyErr_Print();
         LOG_ERROR("isEnabledFor is not callable");
+        Py_DecRef(filter_func);
         goto cleanup;
     }
 
-    set_log_filter(simlog_func);
+    set_log_filter(filter_func);    // Note: This function steals a reference to filter_func.
 
-    argv_list = PyList_New(0);
+    // Build argv for cocotb module
+    argv_list = PyList_New(info->argc);                                     // New reference
     for (i = 0; i < info->argc; i++) {
-        arg_value = PyString_FromString(info->argv[i]);
-        PyList_Append(argv_list, arg_value);
+        PyList_SetItem(argv_list, i, PyString_FromString(info->argv[i]));   // Note: This function steals the reference to created PyString
```

Comment:
Strictly you ought to check that PyString_FromString doesn't return NULL here.

Reply:
If it fails for an string in argv, should we put in an empty string, None, or fail init?

Reply:
Fail init. If it does fail, you're not going to be able to recover.

Reply:
Added more error checking for argv creation.

Reply:
Added error checking in more places.